ts/main.ts

```diff
@@ -0,0 +1,64 @@
+interface FormElements extends HTMLFormControlsCollection {
+  title: HTMLInputElement;
+  url: HTMLInputElement;
+  notes: HTMLTextAreaElement;
+}
+
+const $photoURL = document.querySelector('#photo-url');
+const $photo = document.querySelector('img');
+const $form = document.querySelector('form');
+
+if (!$photoURL) throw new Error('$photoURL query failed');
+if (!$photo) throw new Error('$photo query failed');
+if (!$form) throw new Error('$form query failed');
+
+$photoURL.addEventListener('input', (event: Event) => {
+  const $eventTarget = event.target as HTMLInputElement;
+  const url = $eventTarget.value;
+
+  if (isValid(url)) {
+    $photo.setAttribute('src', url);
+  } else {
+    $photo.setAttribute('src', 'images/placeholder-image-square.jpg');
+  }
+});
+
+$photo.addEventListener('error', () => {
+  $photo.setAttribute('src', 'images/placeholder-image-square.jpg');
+});
+
+$form.addEventListener('submit', (event: Event) => {
+  event.preventDefault();
+
+  if (!$form) return;
+  const $formElements = $form.elements as FormElements;
+
+  const title = $formElements.title.value;
+  const url = $formElements.url.value;
+  const notes = $formElements.notes.value;
+  const entryId = data.nextEntryId;
+
+  const entry: Entry = {
+    title,
+    url,
+    notes,
+    entryId,
+  };
+
+  data.nextEntryId++;
+  data.entries.push(entry);
```

Comment:
`entry` should be getting added to the beginning of the array, rather than the end